### PR TITLE
feat: add dynamic routing header generation

### PR DESCRIPTION
**Pull request by alicejli**
Add support for the [google.api.routing](https://github.com/googleapis/googleapis/blob/master/google/api/routing.proto) annotation. In short, this annotation allows API producers to explicitly define those request fields should be include in request headers/metadata. Furthermore, it enables API producers to define simple segment matchers (primarily for resource name values) that are generated into regular expressions in the client that extract segments of a field's value to be sent in the header. Finally, it enables API producers to specify a key other than the field name to use in the header key-value-pair. In relation to the implicit headers (extraced from google.api.http annotations), the google.api.routing annotation takes precedence. Please read the annotation documentation for more information on the annotation resolution logic.

- This includes a `gapic-showcase` integration test

### CODE REVIEW

File: internal/gengapic/testdata/method_GetAnotherThing.want [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	headerName2 := "foo_name"
+	headerValue2 := ""
+	if routingRegexp2.MatchString(req.GetAnother()){headerValue2 = url.QueryEscape(routingRegexp2.FindStringSubmatch(req.GetAnother())[1])}
+	if len(headerValue2) > 0{routingHeadersMap[headerName2] = headerValue2}
+	routingRegexp3 := regexp.MustCompile("(?P<foo_name>projects/[^/]+/foos/[^/]+)/bars/[^/]+(?:/.*)?")
+	headerName3 := "foo_name"
+	headerValue3 := ""
+	if routingRegexp3.MatchString(req.GetAnother()){headerValue3 = url.QueryEscape(routingRegexp3.FindStringSubmatch(req.GetAnother())[1])}
+	if len(headerValue3) > 0{routingHeadersMap[headerName3] = headerValue3}
+	routingRegexp4 := regexp.MustCompile("(?P<nested_name>.*)")
```
Comment (noahdietz) [inline internal/gengapic/testdata/method_GetAnotherThing.want:29]:
```suggestion
	if reg := regexp.MustCompile("(?P<foo_name>projects/[^/]+)/bars/[^/]+(?:/.*)?"); reg.MatchString(req.GetAnother()) {
	    routingHeadersMap["foo_name"] = url.QueryEscape(reg.FindStringSubmatch(req.GetAnother())[1])
	} else if reg = regexp.MustCompile("(?P<foo_name>projects/[^/]+/foos/[^/]+)/bars/[^/]+(?:/.*)?"); reg.MatchString(req.GetAnother()) {
	    routingHeadersMap["foo_name"] = url.QueryEscape(reg.FindStringSubmatch(req.GetAnother())[1])
	}
```

Here is another idea of how we can structure the generated code taking advantage of Go's ability to define local variables in an `if` and avoiding unnecessary creation of variables for things that the generator already knows about i.e. the header name (we can just use the raw value). I think this greatly improves readability.

File: internal/gengapic/gengapic.go [hunk 2]
```diff
@@ -243,12 +243,49 @@ func (g *generator) deadline(s, m string) {
...
+		// nested enum, the name is prefixed with the parent message name.
+		// For example, Severity_name or Error_Severity_name.
+		value = fmt.Sprintf("%s.%s_name[int32(%s)]", imp.Name, n, accessor)
+	default:
+		value = accessor
+	}
+	return value, nil
+}
+
+func (g *generator) insertRegularRequestHeaders(m *descriptor.MethodDescriptorProto) error {
```
Comment (noahdietz) [inline internal/gengapic/gengapic.go:284]: Could we rename this to `insertImplicitRequestHeaders`? There is no "regular" in our world just one or the other :)

**Review by noahdietz — CHANGES_REQUESTED** [review]

**Comment by noahdietz** [comment]
Great great start to this Alice! This is a tough feature and you are integrating it marvelously.

File: internal/gengapic/helpers.go [hunk 3]
```diff
@@ -162,3 +163,48 @@ func isRequired(field *descriptor.FieldDescriptorProto) bool {
...
+	// Path template should only contain one name (e.g. at most one `=`) or
+	// a collectionId, and should be contained within curly braces.
+	if strings.Count(pattern, "=") > 1 || !curlyBraceRegex.MatchString(pattern) {
+		return ""
+	} else if strings.Count(pattern, "=") < 1 {
+		// If there is no equal sign, then the path template is a collectionId which is its own name.
+		// and both the named segment and path template are wildcards.
+		return curlyBraceRegex.FindStringSubmatch(pattern)[1]
+	}
+	helperSegment := curlyBraceRegex.FindStringSubmatch(pattern)[1]
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/helpers.go]: Renamed this to `curlyBraceSegment`, added a comment and edited the function slightly so it's used in the function. Basically it returns the text within the curly braces of a path template.

File: internal/gengapic/helpers.go [hunk 4]
```diff
@@ -162,3 +163,48 @@ func isRequired(field *descriptor.FieldDescriptorProto) bool {
...
+	curlyBraceRegex := regexp.MustCompile("\\{([^}]+)\\}")
+	// Path template should only contain one name (e.g. at most one `=`) or
+	// a collectionId, and should be contained within curly braces.
+	if strings.Count(pattern, "=") > 1 || !curlyBraceRegex.MatchString(pattern) {
+		return ""
+	} else if strings.Count(pattern, "=") < 1 {
+		// If there is no equal sign, then the path template is a collectionId which is its own name.
+		// and both the named segment and path template are wildcards.
+		return curlyBraceRegex.FindStringSubmatch(pattern)[1]
+	}
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/helpers.go]: done.

File: internal/gengapic/helpers.go [hunk 5]
```diff
@@ -162,3 +163,48 @@ func isRequired(field *descriptor.FieldDescriptorProto) bool {
...
 	return false
 }
+
+// This takes in a path template from a routing annotation and converts it into a regex string.
+// The named capture is the named segment portion for the header itself.
+func convertPathTemplateToRegex(pattern string) string {
+	// If path template doesn't exist, then use a wildcard.
+	if len(pattern) < 1 {
+		return "(.*)"
+	}
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/helpers.go]: done.

File: internal/gengapic/helpers_test.go [hunk 6]
```diff
@@ -15,6 +15,7 @@
 package gengapic
 
 import (
+	//"reflect"
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/helpers_test.go]: done.

File: internal/gengapic/helpers_test.go [hunk 7]
```diff
@@ -192,3 +193,83 @@ func TestIsRequired(t *testing.T) {
 		}
 	}
 }
+
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/helpers_test.go:195]: :D done.

File: internal/gengapic/helpers.go [hunk 8]
```diff
@@ -162,3 +163,48 @@ func isRequired(field *descriptor.FieldDescriptorProto) bool {
...
+	// Replace name of header to named capture.
+	regexPattern := strings.ReplaceAll(pattern, "{", "(?P<")
+	regexPattern = strings.ReplaceAll(regexPattern, "}", ")")
+	// If not named, then entire segment is a wildcard.
+	if !strings.Contains(pattern, "=") || !strings.Contains(pattern, "/") {
+		regexPattern = strings.ReplaceAll(regexPattern, "*", "")
+		regexPattern = strings.ReplaceAll(regexPattern, "=", "")
+		regexPattern = strings.ReplaceAll(regexPattern, ")", ">.*)")
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/helpers.go:181]: Yup! Since this block of code only runs if the entire segment is a wildcard, line 181 will just replace the `)` to convert the entire thing to the named capture.

File: internal/gengapic/helpers.go [hunk 9]
```diff
@@ -162,3 +163,48 @@ func isRequired(field *descriptor.FieldDescriptorProto) bool {
...
+	// a collectionId, and should be contained within curly braces.
+	if strings.Count(pattern, "=") > 1 || !curlyBraceRegex.MatchString(pattern) {
+		return ""
+	} else if strings.Count(pattern, "=") < 1 {
+		// If there is no equal sign, then the path template is a collectionId which is its own name.
+		// and both the named segment and path template are wildcards.
+		return curlyBraceRegex.FindStringSubmatch(pattern)[1]
+	}
+	helperSegment := curlyBraceRegex.FindStringSubmatch(pattern)[1]
+	getBeforeAfterEqualsSign := regexp.MustCompile("(?P<before>[^=]*)=(?P<after>.*)")
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/helpers.go]: it's...not necessary :)  updated!

File: internal/gengapic/helpers_test.go [hunk 10]
```diff
@@ -192,3 +193,83 @@ func TestIsRequired(t *testing.T) {
...
+			in:   "{foo=*}",
+			want: "(?P<foo>.*)",
+		},
+		{
+			in:   "{foo=**}",
+			want: "(?P<foo>.*)",
+		},
+		{
+			in:   "{foo=projects/*}/bars",
+			want: "(?P<foo>projects/[^/]+)/bars",
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/helpers_test.go:219]: Nope - forward slashes are not considered special characters. Confirmed on both the Go playground + regex101.com

File: internal/gengapic/gengapic.go [hunk 11]
```diff
@@ -531,3 +590,38 @@ func parseRequestHeaders(m *descriptor.MethodDescriptorProto) ([][]string, error
...
+}
+
+// Parse routing annotations to be used as request headers
+func parseDynamicRequestHeaders(m *descriptor.MethodDescriptorProto) ([][]string, error) {
+	var matches [][]string
+
+	reqHeaders := proto.GetExtension(m.GetOptions(), annotations.E_Routing)
+	if m == nil || m.GetOptions() == nil {
+		return nil, nil
+	}
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/gengapic.go]: Ah thank you for explaining! That makes sense - removed.

File: internal/gengapic/gengapic.go [hunk 12]
```diff
@@ -531,3 +590,38 @@ func parseRequestHeaders(m *descriptor.MethodDescriptorProto) ([][]string, error
...
 	return matches, nil
 }
+
+// Determine whether routing annotation exists
+func dynamicRequestHeadersExist(m *descriptor.MethodDescriptorProto) bool {
+	dynamicHeaders := proto.HasExtension(m.GetOptions(), annotations.E_Routing)
+	if m == nil || m.GetOptions() == nil {
+		return false
+	}
+	return dynamicHeaders
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/gengapic.go]: `dynamicHeaders` should be a `bool` (is `HasExtension` rather than `GetExtension`). 
Good call - I added a test for `dynamicRequestHeadersExist` (returns `true` if extension is set but empty, which is intended).

File: internal/gengapic/gengapic.go [hunk 13]
```diff
@@ -259,62 +296,84 @@ func (g *generator) insertMetadata(m *descriptor.MethodDescriptorProto) error {
...
+		g.insertRegularRequestHeaders(m)
+		g.printf("ctx = insertMetadata(ctx, c.xGoogMetadata, md)")
+		g.imports[pbinfo.ImportSpec{Path: "fmt"}] = true
+		g.imports[pbinfo.ImportSpec{Path: "net/url"}] = true
+		return nil
+	} else if len(headers) > 0 && dynamicRequestHeadersExist(m) {
+		g.insertDynamicRequestHeaders(m)
+		g.printf("ctx = insertMetadata(ctx, c.xGoogMetadata, md)")
 		g.imports[pbinfo.ImportSpec{Path: "fmt"}] = true
 		g.imports[pbinfo.ImportSpec{Path: "net/url"}] = true
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/gengapic.go]: thank you; done!

File: internal/gengapic/gengapic.go [hunk 14]
```diff
@@ -259,62 +296,84 @@ func (g *generator) insertMetadata(m *descriptor.MethodDescriptorProto) error {
...
+	return nil
+}
+
+func (g *generator) insertDynamicRequestHeaders(m *descriptor.MethodDescriptorProto) error {
+	headers, err := parseDynamicRequestHeaders(m)
+	if err != nil {
+		return err
+	}
+	if len(headers) > 0 {
+		g.printf("routingHeaders := \"\"")
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/gengapic.go]: done

File: internal/gengapic/gengapic.go [hunk 15]
```diff
@@ -259,62 +296,84 @@ func (g *generator) insertMetadata(m *descriptor.MethodDescriptorProto) error {
...
+			accessor := fmt.Sprintf("req%s", fieldGetter(field))
+			regexHelper := fmt.Sprintf("routingRegexp%v.FindStringSubmatch(%s)[1]", i, accessor)
+			g.printf("headerName%v := \"%s\"", i, headerName)
+			g.printf("headerValue%v := \"\"", i)
+
+			regexHelper, err = g.getFormattedValue(m, field, regexHelper)
+			if err != nil {
+				return err
+			}
+			g.printf("if routingRegexp%v.MatchString(%s){headerValue%v = %s}", i, accessor, i, regexHelper)
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/gengapic.go]: thank you - makes it much more readable :) done.

File: internal/gengapic/gengapic.go [hunk 16]
```diff
@@ -259,62 +296,84 @@ func (g *generator) insertMetadata(m *descriptor.MethodDescriptorProto) error {
...
+			// This enables overwriting of headerNames such that the last annotation to match will be the value sent.
+			g.printf("if len(headerValue%v) > 0{routingHeadersMap[headerName%v] = headerValue%v}", i, i, i)
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/gengapic.go]: done

File: internal/gengapic/gengapic.go [hunk 17]
```diff
@@ -259,62 +296,84 @@ func (g *generator) insertMetadata(m *descriptor.MethodDescriptorProto) error {
...
+		}
+		g.printf("for headerName, headerValue := range routingHeadersMap {routingHeaders = routingHeaders + fmt.Sprintf(\"%%s=%%s&\", headerName, headerValue)}")
+		// Trim the trailing ampersand symbol.
+		g.printf("if len(routingHeaders) > 0{routingHeaders = routingHeaders[:len(routingHeaders)-1]}")
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/gengapic.go]: thank you - updated!

File: internal/gengapic/testdata/method_GetAnotherThing.want (hunk 1 above)
Comment (alicejli) [reply internal/gengapic/testdata/method_GetAnotherThing.want]: that's really neat; thank you for explaining! I've refactored `insertDynamicRequestHeaders` to match the above.

File: internal/gengapic/gengapic.go (hunk 2 above)
Comment (alicejli) [reply internal/gengapic/gengapic.go]: ha - makes sense. Done!

File: internal/gengapic/gengapic.go [hunk 18]
```diff
@@ -318,56 +318,70 @@ func (g *generator) insertDynamicRequestHeaders(m *descriptor.MethodDescriptorPr
...
+				g.printf("}")
+			} else {
+				g.printf("}")
+			}
 		}
-		g.printf("for headerName, headerValue := range routingHeadersMap {routingHeaders = routingHeaders + fmt.Sprintf(\"%%s=%%s&\", headerName, headerValue)}")
-		// Trim the trailing ampersand symbol.
-		g.printf("if len(routingHeaders) > 0{routingHeaders = routingHeaders[:len(routingHeaders)-1]}")
+		g.printf("for headerName, headerValue := range routingHeadersMap {")
+		g.printf("routingHeaders = routingHeaders + fmt.Sprintf(\"%%s=%%s&\", headerName, headerValue)")
```
Comment (noahdietz) [inline internal/gengapic/gengapic.go:355]: indent, use backtick so you don't need to escape the double quotes, concat with `routingHeaders` by including it in the format string with leading `%s`
```suggestion
		g.printf(`  routingHeaders = fmt.Sprintf("%%s%%s=%%s&", routingHeaders, headerName, headerValue)`)
```

File: internal/gengapic/gengapic.go [hunk 19]
```diff
@@ -318,56 +318,70 @@ func (g *generator) insertDynamicRequestHeaders(m *descriptor.MethodDescriptorPr
...
+				namedCaptureRegex := headers[i+1][0]
+				field := headers[i+1][1]
+				accessor := fmt.Sprintf("req%s", fieldGetter(field))
+				regexHelper := fmt.Sprintf("reg.FindStringSubmatch(%s)[1]", accessor)
+				regexHelper, err = g.getFormattedValue(m, field, regexHelper)
+				if err != nil {
+					return err
+				}
+				g.printf("} else if reg = regexp.MustCompile(\"%s\");", namedCaptureRegex)
+				g.printf("reg.MatchString(%s) {", accessor)
```
Comment (noahdietz) [inline internal/gengapic/gengapic.go:347]: Same line, use `%q`
```suggestion
				g.printf("} else if reg = regexp.MustCompile(%q); reg.MatchString(%s) {", namedCaptureRegex, accessor)
```

File: internal/gengapic/gengapic.go [hunk 20]
```diff
@@ -318,56 +318,70 @@ func (g *generator) insertDynamicRequestHeaders(m *descriptor.MethodDescriptorPr
...
+				g.printf("routingHeadersMap[\"%s\"] = %s", headerName, regexHelper)
```
Comment (noahdietz) [inline internal/gengapic/gengapic.go:348]: Indent and use `%q`
```suggestion
				g.printf("  routingHeadersMap[%q] = %s", headerName, regexHelper)
```

File: internal/gengapic/gengapic.go [hunk 21]
```diff
@@ -593,22 +607,15 @@ func parseRequestHeaders(m *descriptor.MethodDescriptorProto) ([][]string, error
 
 // Determine whether routing annotation exists
 func dynamicRequestHeadersExist(m *descriptor.MethodDescriptorProto) bool {
-	dynamicHeaders := proto.HasExtension(m.GetOptions(), annotations.E_Routing)
-	if m == nil || m.GetOptions() == nil {
-		return false
-	}
-	return dynamicHeaders
+	dynamicHeadersExt := proto.HasExtension(m.GetOptions(), annotations.E_Routing)
+	return dynamicHeadersExt
```
Comment (noahdietz) [inline internal/gengapic/gengapic.go:611]:
```suggestion
	return proto.HasExtension(m.GetOptions(), annotations.E_Routing)
```

File: internal/gengapic/testdata/method_GetAnotherThing.want [hunk 22]
```diff
@@ -6,39 +6,39 @@ func (c *fooGRPCClient) GetAnotherThing(ctx context.Context, req *mypackagepb.In
...
+	reg.MatchString(req.GetAnother()) {
+		routingHeadersMap["foo_name"] = url.QueryEscape(reg.FindStringSubmatch(req.GetAnother())[1])
+	} else if reg = regexp.MustCompile("(?P<foo_name>projects/[^/]+/foos/[^/]+)/bars/[^/]+(?:/.*)?");
+	reg.MatchString(req.GetAnother()) {
+		routingHeadersMap["foo_name"] = url.QueryEscape(reg.FindStringSubmatch(req.GetAnother())[1])
+	}
+	if reg := regexp.MustCompile("(?P<foo_name>projects/[^/]+/foos/[^/]+)/bars/[^/]+(?:/.*)?");
+	reg.MatchString(req.GetAnother()) {
+		routingHeadersMap["foo_name"] = url.QueryEscape(reg.FindStringSubmatch(req.GetAnother())[1])
+	}
```
Comment (noahdietz) [inline internal/gengapic/testdata/method_GetAnotherThing.want:27]: I think this should be an `else if` connected to the `if`/`else if` immediately above, yeah?

**Review by noahdietz — CHANGES_REQUESTED** [review]
Great adjustments.

File: internal/gengapic/gengapic.go [hunk 23]
```diff
@@ -318,56 +318,70 @@ func (g *generator) insertDynamicRequestHeaders(m *descriptor.MethodDescriptorPr
...
+			regexHelper := fmt.Sprintf("reg.FindStringSubmatch(%s)[1]", accessor)
 			regexHelper, err = g.getFormattedValue(m, field, regexHelper)
 			if err != nil {
 				return err
 			}
-			g.printf("if routingRegexp%v.MatchString(%s){headerValue%v = %s}", i, accessor, i, regexHelper)
-			// This enables overwriting of headerNames such that the last annotation to match will be the value sent.
-			g.printf("if len(headerValue%v) > 0{routingHeadersMap[headerName%v] = headerValue%v}", i, i, i)
+			g.printf("if reg := regexp.MustCompile(\"%s\");", namedCaptureRegex)
+			g.printf("reg.MatchString(%s) {", accessor)
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/gengapic.go:334]: oh nice! Thanks - done.

File: internal/gengapic/gengapic.go [hunk 24]
```diff
@@ -318,56 +318,70 @@ func (g *generator) insertDynamicRequestHeaders(m *descriptor.MethodDescriptorPr
...
+			g.printf("routingHeadersMap[\"%s\"] = %s", headerName, regexHelper)
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/gengapic.go:335]: makes sense - done.

File: internal/gengapic/gengapic.go [hunk 25]
```diff
@@ -318,56 +318,70 @@ func (g *generator) insertDynamicRequestHeaders(m *descriptor.MethodDescriptorPr
...
+				if err != nil {
+					return err
+				}
+				g.printf("} else if reg = regexp.MustCompile(\"%s\");", namedCaptureRegex)
+				g.printf("reg.MatchString(%s) {", accessor)
+				g.printf("routingHeadersMap[\"%s\"] = %s", headerName, regexHelper)
+				g.printf("}")
+			} else {
+				g.printf("}")
+			}
```
[the remark this answers is not in the document]
Comment (alicejli) [reply internal/gengapic/gengapic.go:352]: done

File: internal/gengapic/gengapic.go (hunk 18 above)
Comment (alicejli) [reply internal/gengapic/gengapic.go:355]: done

File: internal/gengapic/gengapic.go (hunk 19 above)
Comment (alicejli) [reply internal/gengapic/gengapic.go:347]: done

File: internal/gengapic/doc_file.go [hunk 26]
```diff
@@ -111,6 +111,7 @@ func (g *generator) genDocFile(year int, scopes []string, serv *descriptor.Servi
 	p("import (")
 	p("%s%q", "\t", "context")
 	if hasREST {
+		p("%s%q", "\t", "fmt")
```
Comment (noahdietz) [inline internal/gengapic/doc_file.go:114]: I think your rebase got kind of messed up because this was already commited and shouldn't show up in the diff

**Review by noahdietz — COMMENTED** [review]
This is looking really good!

File: internal/gengapic/testdata/method_GetAnotherThing.want (hunk 22 above)
Comment (noahdietz) [reply internal/gengapic/testdata/method_GetAnotherThing.want]: I think just straight `if` statements might be better actually, so let's stick with what you have.

File: internal/gengapic/doc_file.go (hunk 26 above)
Comment (alicejli) [reply internal/gengapic/doc_file.go]: holy cow I'm sorry my git got totally messed up. I think it's fixed now, and only files that should be getting changed are actually being changed in this PR. I'm sorry >.<

File: showcase/echo_test.go [hunk 27]
```diff
@@ -94,6 +97,63 @@ func TestEcho_error(t *testing.T) {
 
 }
 
+// Test dynamic routing header generation.
+func TestEchoHeader(t *testing.T) {
+	defer check(t)
+	content := "hello world!"
+	reqs := []*showcasepb.EchoRequest{&showcasepb.EchoRequest{
```
Comment (noahdietz) [inline showcase/echo_test.go:104]: Take a look at this piece about [Table Driven Tests](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) and look elsewhere in our codebase. Let's refactor this list of requests into a list of test cases with the expected headers included so we can say "for this request, I expect this header".

Comment (alicejli) [reply showcase/echo_test.go]: Thanks for sharing! Makes sense. I've refactored the test a bit (it also helped demonstrate an edge case that I've updated `insertDynamicRequestHeaders` to manage). Let me know how it looks now.

File: showcase/echo_test.go [hunk 28]
```diff
@@ -94,6 +99,50 @@ func TestEcho_error(t *testing.T) {
...
+			want: []string{"baz=projects%2F123%2Finstances%2F456", "qux=projects%2F123", "table_name=regions%2F123%2Fzones%2F456", "header=regions%2F123%2Fzones%2F456", "routing_id=regions%2F123%2Fzones%2F456"},
+		},
+	} {
+		mdForHeaders := metadata.New(map[string]string{})
+		echo.Echo(context.Background(), tst.req, gax.WithGRPCOptions(grpc.Header(&mdForHeaders)))
+		got := mdForHeaders.Get("x-goog-request-params")
+		got = strings.Split(got[0], "&")
+		sort.Strings(got)
+		sort.Strings(tst.want)
+		if !reflect.DeepEqual(got, tst.want) {
```
Comment (noahdietz) [inline showcase/echo_test.go:140]: Nit: use cmp.Diff (look around the codebase for uses) to compare to values in a test. This can be done in another PR though.

**Review by noahdietz — APPROVED** [review]
Way to go, Alice, well done.
